src/prefect/cli/agent.py

```diff
@@ -97,15 +101,82 @@ async def start(
         if not hide_welcome:
             app.console.print(ascii_name)
             app.console.print(
-                f"Agent started! Looking for work from queue '{work_queue_name or work_queue_id}'..."
+                "Agent started! Looking for work from "
+                f"queue '{work_queue_name or work_queue_id}'..."
             )
 
-        while True:
-            try:
-                await agent.get_and_submit_flow_runs()
-            except KeyboardInterrupt:
-                break
-
-            await anyio.sleep(PREFECT_AGENT_QUERY_INTERVAL.value())
+        await critical_service_loop(
+            agent.get_and_submit_flow_runs,
+            PREFECT_AGENT_QUERY_INTERVAL.value(),
+        )
 
     app.console.print("Agent stopped!")
+
+
+async def critical_service_loop(
+    workload: Callable[..., Coroutine],
+    interval: float,
+    memory: int = 20,
+):
+    """
+    Runs the the given `workload` function on the specified `interval`, while being
+    forgiving of intermittent issues like temporary HTTP errors.  If more failures than
+    successes occur, prints a summary of recently seen errors, then returns.
+
+    Args:
+        `workload`: the function to call
+        `interval`: how frequently to call it
+        `memory`: how many successes/failures to remember when deciding whether to exit
+    """
+
+    track_record: Deque[int] = deque(maxlen=memory)
+    exceptions: Deque[Optional[Exception]] = deque(maxlen=track_record.maxlen)
+
+    while True:
+        try:
+            await workload()
+
+            track_record.append(1)
+            exceptions.append(None)
+        except httpx.TransportError as exc:
+            # httpx.TransportError is the base class for any kind of communications
+            # error, like timeouts, connection failures, etc.  This does _not_ cover
+            # routine HTTP error codes (even 5xx errors like 502/503) so this
+            # handler should not be attempting to cover cases where the Orion server
+            # or Prefect Cloud is having an outage (which will be covered by the
+            # exception clause below)
+            track_record.append(-1)
+            exceptions.append(exc)
+        except httpx.HTTPStatusError as exc:
+            if exc.response.status_code in (502, 503):
+                # 502/503 indicate a potential outage of the Orion server or Prefect
+                # Cloud, which is likely to be temporary and transient.  Don't quit
+                # over these unless it is prolonged.
+                track_record.append(-1)
+                exceptions.append(exc)
+            else:
+                raise
+        except KeyboardInterrupt:
+            return
+
+        # If the last result was a failure, see if it's time to cut our losses because
+        # we're failing more often than we're succeeding
+        if track_record[-1] < 0 and sum(track_record) < 0:
```

Comment:
`sum(track_record)` is essentially saying "more than half", which is probably something we'd like to tune?